api/buckets/ipnsmanager.go

```diff
@@ -0,0 +1,102 @@
+package buckets
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	iface "github.com/ipfs/interface-go-ipfs-core"
+	opt "github.com/ipfs/interface-go-ipfs-core/options"
+	"github.com/ipfs/interface-go-ipfs-core/path"
+	c "github.com/textileio/textile/collections"
+)
+
+// IPNSManager handles bucket name publishing to IPNS.
+type IPNSManager struct {
+	keys     *c.IPNSKeys
+	api      iface.NameAPI
+	lock     sync.Mutex
+	locks    map[string]chan struct{}
+	ctxsLock sync.Mutex
+	ctxs     map[string]context.CancelFunc
+}
+
+// NewIPNSManager returns a new IPNS manager.
+func NewIPNSManager(keys *c.IPNSKeys, api iface.NameAPI) *IPNSManager {
+	return &IPNSManager{
+		keys:  keys,
+		api:   api,
+		ctxs:  make(map[string]context.CancelFunc),
+		locks: make(map[string]chan struct{}),
+	}
+}
+
+// Cancel all pending publishes.
+func (m *IPNSManager) Cancel() {
+	m.lock.Unlock()
+	defer m.lock.Unlock()
+	for _, cancel := range m.ctxs {
+		cancel()
+	}
+}
+
+func (m *IPNSManager) publish(pth path.Path, keyID string) {
```

Comment:
Some points about the goals here:
- Only one publish per key at a time
- We only want to most recent path to be published
- If a new publish request comes in, cancel the pending one (for example, when pushing a directory of files)